src/Logging/Logging.AzureAppServices/src/AzureAppServicesLoggerFactoryExtensions.cs

```diff
@@ -0,0 +1,98 @@
+        /// </summary>
+        /// <param name="builder">The extension method argument</param>
+        public static ILoggingBuilder AddAzureWebAppDiagnostics(this ILoggingBuilder builder)
+        {
+            var context = WebAppContext.Default;
+
+            // Only add the provider if we're in Azure WebApp. That cannot change once the apps started
+            return AddAzureWebAppDiagnostics(builder, context);
+        }
+
+        internal static ILoggingBuilder AddAzureWebAppDiagnostics(this ILoggingBuilder builder, IWebAppContext context)
+        {
+            if (!context.IsRunningInAzureWebApp)
+            {
+                return builder;
+            }
+
+            builder.AddConfiguration();
+
+            var config = SiteConfigurationProvider.GetAzureLoggingConfiguration(context);
+            var services = builder.Services;
+
+            var addedFileLogger = TryAddEnumerable(services, Singleton<ILoggerProvider, FileLoggerProvider>());
+            var addedBlobLogger = TryAddEnumerable(services, Singleton<ILoggerProvider, BlobLoggerProvider>());
+
+            if (addedFileLogger || addedBlobLogger)
+            {
+                services.AddSingleton(context);
+                services.AddSingleton<IOptionsChangeTokenSource<LoggerFilterOptions>>(
+                    new ConfigurationChangeTokenSource<LoggerFilterOptions>(config));
+            }
+
+            if (addedFileLogger)
+            {
+                services.AddSingleton<IConfigureOptions<LoggerFilterOptions>>(CreateFileFilterConfigureOptions(config));
+                services.AddSingleton<IConfigureOptions<AzureFileLoggerOptions>>(new FileLoggerConfigureOptions(config, context));
+                services.AddSingleton<IOptionsChangeTokenSource<AzureFileLoggerOptions>>(
+                    new ConfigurationChangeTokenSource<AzureFileLoggerOptions>(config));
+                LoggerProviderOptions.RegisterProviderOptions<AzureFileLoggerOptions, FileLoggerProvider>(builder.Services);
+            }
+
+            if (addedBlobLogger)
+            {
+                services.AddSingleton<IConfigureOptions<LoggerFilterOptions>>(CreateBlobFilterConfigureOptions(config));
+                services.AddSingleton<IConfigureOptions<AzureBlobLoggerOptions>>(new BlobLoggerConfigureOptions(config, context));
+                services.AddSingleton<IOptionsChangeTokenSource<AzureBlobLoggerOptions>>(
+                    new ConfigurationChangeTokenSource<AzureBlobLoggerOptions>(config));
+                LoggerProviderOptions.RegisterProviderOptions<AzureBlobLoggerOptions, BlobLoggerProvider>(builder.Services);
+            }
+
+            return builder;
+        }
+
+        private static bool TryAddEnumerable(IServiceCollection collection, ServiceDescriptor descriptor)
+        {
+            var beforeCount = collection.Count;
+            collection.TryAddEnumerable(descriptor);
+            return beforeCount != collection.Count;
```

Comment:
Since other threads can change the `Count` of this shared collection, suggest instead doing something like what `TryAddEnumerable(...)` does:
``` c#
var implementationType = descriptor.GetImplementationType();
if (!services.Any(d =>
    d.ServiceType == descriptor.ServiceType &&
    d.GetImplementationType() == implementationType))
{
    collection.TryAddEnumerable(descriptor);
    return true;
}

return false;
```

Reply:
Can you file a bug for this? I'm concerned with migrating files here, I'm not going to fix bugs as part of this PR.